src/cascade/model/grids.py

```diff
@@ -0,0 +1,174 @@
+            time_step: the spacing of time points
+        """
+        ages = list(range(age_start, age_end, age_step))
+        times = list(range(time_start, time_end, time_step))
+
+        return cls(ages, times)
+
+    def __init__(self, ages, times, snap_distance=GRID_SNAP_DISTANCE):
+        self._ages = tuple(sorted(set(ages)))
+        self._times = tuple(sorted(set(times)))
+        self._snap_distance = snap_distance
+
+    @property
+    def ages(self):
+        return self._ages
+
+    @property
+    def times(self):
+        return self._times
+
+    @property
+    def snap_distance(self):
+        return self._snap_distance
+
+
+def _any_close(value, targets, tolerance):
+    """True if any element of targets is within tolerance of value
+    """
+    for target in targets:
+        if isclose(value, target, abs_tol=tolerance):
+            return True
+    return False
+
+
+def _validate_region(grid, lower_age, upper_age, lower_time, upper_time):
+    """Validates that the described rectangle aligns with the grid.
+    Positive or negative infinite bounds will be treated as stretching to the edge of the grid.
+
+    Raises:
+        ValueError: If the rectangle does not align
+    """
+    if not _any_close(lower_age, grid.ages, grid.snap_distance) and not isinf(lower_age):
+        raise ValueError("Lower age not in underlying grid")
+
+    if not _any_close(upper_age, grid.ages, grid.snap_distance) and not isinf(upper_age):
+        raise ValueError("Upper age not in underlying grid")
+
+    if not _any_close(lower_time, grid.times, grid.snap_distance) and not isinf(lower_time):
+        raise ValueError("Lower time not in underlying grid")
+
+    if not _any_close(upper_time, grid.times, grid.snap_distance) and not isinf(upper_time):
+        raise ValueError("Upper time not in underlying grid")
+
+
+class _RegionView:
+    def __init__(self, parent, age_slice, time_slice):
+        self._parent = parent
+        self._age_slice = age_slice
+        self._time_slice = time_slice
+
+    def _rectangle(self):
+        if self._age_slice.start is None:
+            lower_age = float("-inf")
+        else:
+            lower_age = self._age_slice.start
+
+        if self._age_slice.stop is None:
+            upper_age = float("inf")
+        else:
+            upper_age = self._age_slice.stop
+
+        if self._time_slice.start is None:
+            lower_time = float("-inf")
+        else:
+            lower_time = self._time_slice.start
+
+        if self._time_slice.stop is None:
+            upper_time = float("inf")
+        else:
+            upper_time = self._time_slice.stop
+
+        return (lower_age, upper_age, lower_time, upper_time)
+
+    @property
+    def prior(self):
+        lower_age, upper_age, lower_time, upper_time = self._rectangle()
+        if lower_age != upper_age or lower_time != upper_time:
+            raise NotImplementedError("Currently only point queries are supported")
```

Comment:
Looks like a point means a single age OR a single time, just checking, is that the intention; i.e. doesn't have to be both a single age and a single time?

Reply:
sure, thanks.  I am still learning my approach to PR reading.  OR is complicated.  :-)